hack/test-go.sh

```diff
@@ -1,125 +1,231 @@
+            # we do it ourselves
+            subpackages="$( list_test_packages_under "${package%"/..."}" )"
+            for subpackage in "${subpackages}"; do
+                test_packages="${test_packages} ${subpackage}"
+            done
+        else
+            test_packages="${test_packages} ${OS_GO_PACKAGE}/${package}"
+        fi
+    done
+else
+    # If no packages are given to test, we need to generate a list of all packages with unit tests
+    openshift_test_packages="$(list_test_packages_under '*')"  
 
-  echo 'mode: atomic' > ${OUTPUT_COVERAGE}/profiles.out
-  find $OUTPUT_COVERAGE -name profile.out | xargs sed '/^mode: atomic$/d' >> ${OUTPUT_COVERAGE}/profiles.out
-  go tool cover "-html=${OUTPUT_COVERAGE}/profiles.out" -o "${OUTPUT_COVERAGE}/coverage.html"
+    kubernetes_path="Godeps/_workspace/src/k8s.io/kubernetes"
+    mandatory_kubernetes_packages="$( echo "${OS_GO_PACKAGE}/${kubernetes_path}/pkg/api"; echo "${OS_GO_PACKAGE}/${kubernetes_path}/pkg/api/v1"; )"
 
-  rm -rf $OUTPUT_COVERAGE/$OS_GO_PACKAGE
-else
-  nice go test $KUBE_RACE -timeout $KUBE_TIMEOUT $KUBE_COVER "${@:2}" $test_packages
+    test_packages="$(echo "${openshift_test_packages}"; echo "${mandatory_kubernetes_packages}")"
+
+    if [[ -n "${test_kube}" ]]; then
+        optional_kubernetes_packages="$(find "${kubernetes_path}" -not \( \
+          \( \
+            -wholename "${kubernetes_path}/pkg/api" \
+            -o -wholename "${kubernetes_path}/pkg/api/v1" \
+            -o -wholename "${kubernetes_path}/test/e2e" \
+          \) -prune \
+        \) -name '*_test.go' -print0 | xargs -0n1 dirname | sort -u | xargs -n1 printf "${OS_GO_PACKAGE}/%s\n")"
+
+        test_packages="$(echo "${test_packages}"; echo "${optional_kubernetes_packages}")"
+    fi
 fi
 
-ret=$?; ENDTIME=$(date +%s); echo "$0 took $(($ENDTIME - $STARTTIME)) seconds"; exit "$ret"
+if [[ -n "${print_packages}" ]]; then
+    echo "The following packages will be tested by $0:"
+    for package in "${test_packages}"; do
+        echo "${package}"
+    done
+    exit 0
+fi
+
+# Run 'go test' with the accumulated arguments and packages:
+if [[ -n "${junit_report}" ]]; then
+    # we need to generate jUnit xml
+    hack/build-go.sh tools/junitreport
+    junitreport="$(os::build::find-binary junitreport)"
+
+    if [[ -z "${junitreport}" ]]; then
+        echo "It looks as if you don't have a compiled junitreport binary"
+        echo
+        echo "If you are running from a clone of the git repo, please run"
+        echo "'./hack/build-go.sh tools/junitreport'."
+        exit 1
+    fi
+
+    test_output_file="${LOG_DIR}/test-go.log"
+    junit_report_file="${ARTIFACT_DIR}/report.xml"
+
+    # we don't care if 'go test' fails, we want to summarize the output anyway.
+    set +o errexit
+    go test ${gotest_flags} ${test_packages} > ${test_output_file} 2>&1
+    return_code=$?
+    set -o errexit
```

Comment:
It would be simpler to do the following:
```shell
return_code=0
go test ${gotest_flags} ${test_packages} > ${test_output_file} 2>&1 || return_code=$?
```